File: backend/src/main/java/com/aprendizaje/backend/service/ProductoService.java
package com.aprendizaje.backend.service;

import com.aprendizaje.backend.model.Producto;
import com.aprendizaje.backend.repository.ProductoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Servicio que contiene la lógica de negocio para los Productos
 * @Service indica que esta clase es un servicio de Spring
 */
@Service
public class ProductoService {
    
    /**
     * Inyección de dependencias del repositorio
     * @Autowired indica que Spring debe inyectar automáticamente el repositorio
     */
    @Autowired
    private ProductoRepository productoRepository;
    
    /**
     * Obtiene todos los productos de la base de datos
     * @return lista de todos los productos
     */
    public List<Producto> obtenerTodosLosProductos() {
        // Llama al método findAll() del repositorio
        return productoRepository.findAll();
    }
    
    /**
     * Obtiene un producto por su ID
     * @param id - identificador del producto
     * @return Optional con el producto si existe, vacío si no existe
     */
    public Optional<Producto> obtenerProductoPorId(Long id) {
        // Llama al método findById() del repositorio
        return productoRepository.findById(id);
    }
    
    /**
     * Crea un nuevo producto en la base de datos
     * @param producto - producto a crear
     * @return el producto creado con su ID generado
     */
    public Producto crearProducto(Producto producto) {
        // Llama al método save() del repositorio
        return productoRepository.save(producto);
    }
    
    /**
     * Actualiza un producto existente
     * @param id - identificador del producto a actualizar
     * @param productoActualizado - datos actualizados del producto
     * @return el producto actualizado, o null si no existe
     */
    public Producto actualizarProducto(Long id, Producto productoActualizado) {
        // Verifica si el producto existe
        Optional<Producto> productoExistente = productoRepository.findById(id);
        
        if (productoExistente.isPresent()) {
            // Si existe, actualiza los campos
            Producto producto = productoExistente.get();
            producto.setNombre(productoActualizado.getNombre());
            producto.setDescripcion(productoActualizado.getDescripcion());
            producto.setPrecio(productoActualizado.getPrecio());
            producto.setStock(productoActualizado.getStock());
            producto.setCategoria(productoActualizado.getCategoria());
            
            // Guarda los cambios
            return productoRepository.save(producto);
        }
        
        // Si no existe, retorna null
        return null;
    }
    
    /**
     * Elimina un producto por su ID
     * @param id - identificador del producto a eliminar
     * @return true si se eliminó, false si no existía
     */
    public boolean eliminarProducto(Long id) {
        // Verifica si existe
        if (productoRepository.existsById(id)) {
            // Si existe, lo elimina
            productoRepository.deleteById(id);
            return true;
        }
        // Si no existe, retorna false
        return false;
    }
    
    /**
     * Busca productos por categoría
     * @param categoria - categoría a buscar
     * @return lista de productos de esa categoría
     */
    public List<Producto> buscarPorCategoria(String categoria) {
        // Usa el método personalizado del repositorio
        return productoRepository.findByCategoria(categoria);
    }
    
    /**
     * Busca productos por nombre (búsqueda parcial)
     * @param nombre - texto a buscar en el nombre
     * @return lista de productos que coinciden
     */
    public List<Producto> buscarPorNombre(String nombre) {
        // Usa el método personalizado del repositorio
        return productoRepository.findByNombreContainingIgnoreCase(nombre);
    }
    
    /**
     * Obtiene productos con stock disponible
     * @param stockMinimo - cantidad mínima de stock requerida
     * @return lista de productos con stock suficiente
     */
    public List<Producto> obtenerProductosConStock(Integer stockMinimo) {
        // Usa el método personalizado del repositorio
        return productoRepository.findByStockGreaterThanEqual(stockMinimo);
    }
}

File: frontend/src/services/productoService.js
// URL base de la API
const API_URL = 'http://localhost:8080/api/productos';

/**
 * Servicio para manejar todas las operaciones relacionadas con Productos
 * Centraliza las llamadas HTTP al backend
 */
const productoService = {
  
  /**
   * Obtiene todos los productos del backend
   * @returns {Promise<Array>} Lista de productos
   */
  obtenerTodos: async () => {
    try {
      const respuesta = await fetch(API_URL);
      if (!respuesta.ok) {
        throw new Error('Error al obtener los productos');
      }
      return await respuesta.json();
    } catch (error) {
      console.error('Error en obtenerTodos:', error);
      throw error;
    }
  },

  /**
   * Obtiene un producto por su ID
   * @param {number} id - ID del producto
   * @returns {Promise<Object>} El producto encontrado
   */
  obtenerPorId: async (id) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}`);
      if (!respuesta.ok) {
        throw new Error('Producto no encontrado');
      }
      return await respuesta.json();
    } catch (error) {
      console.error('Error en obtenerPorId:', error);
      throw error;
    }
  },

  /**
   * Crea un nuevo producto
   * @param {Object} producto - Datos del producto a crear
   * @returns {Promise<Object>} El producto creado
   */
  crear: async (producto) => {
    try {
      const respuesta = await fetch(API_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(producto)
      });
      
      if (!respuesta.ok) {
        throw new Error('Error al crear el producto');
      }
      
      return await respuesta.json();
    } catch (error) {
      console.error('Error en crear:', error);
      throw error;
    }
  },

  /**
   * Actualiza un producto existente
   * @param {number} id - ID del producto a actualizar
   * @param {Object} producto - Nuevos datos del producto
   * @returns {Promise<Object>} El producto actualizado
   */
  actualizar: async (id, producto) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(producto)
      });
      
      if (!respuesta.ok) {
        throw new Error('Error al actualizar el producto');
      }
      
      return await respuesta.json();
    } catch (error) {
      console.error('Error en actualizar:', error);
      throw error;
    }
  },

  /**
   * Elimina un producto
   * @param {number} id - ID del producto a eliminar
   * @returns {Promise<boolean>} true si se eliminó correctamente
   */
  eliminar: async (id) => {
    try {
      const respuesta = await fetch(`${API_URL}/${id}`, {
        method: 'DELETE'
      });
      
      if (!respuesta.ok && respuesta.status !== 204) {
        throw new Error('Error al eliminar el producto');
      }
      
      return true;
    } catch (error) {
      console.error('Error en eliminar:', error);
      throw error;
    }
  },

  /**
   * Busca productos por nombre
   * @param {string} nombre - Texto a buscar en el nombre
   * @returns {Promise<Array>} Lista de productos que coinciden
   */
  buscarPorNombre: async (nombre) => {
    try {
      const respuesta = await fetch(`${API_URL}/buscar?nombre=${encodeURIComponent(nombre)}`);
      if (!respuesta.ok) {
        throw new Error('Error al buscar productos');
      }
      return await respuesta.json();
    } catch (error) {
      console.error('Error en buscarPorNombre:', error);
      throw error;
    }
  },

  /**
   * Obtiene productos por categoría
   * @param {string} categoria - Categoría a buscar
   * @returns {Promise<Array>} Lista de productos de esa categoría
   */
  obtenerPorCategoria: async (categoria) => {
    try {
      const respuesta = await fetch(`${API_URL}/categoria/${encodeURIComponent(categoria)}`);
      if (!respuesta.ok) {
        throw new Error('Error al obtener productos por categoría');
      }
      return await respuesta.json();
    } catch (error) {
      console.error('Error en obtenerPorCategoria:', error);
      throw error;
    }
  },

  /**
   * Obtiene productos con stock mínimo
   * @param {number} stockMinimo - Cantidad mínima de stock
   * @returns {Promise<Array>} Lista de productos con stock suficiente
   */
  obtenerConStock: async (stockMinimo) => {
    try {
      const respuesta = await fetch(`${API_URL}/stock/${stockMinimo}`);
      if (!respuesta.ok) {
        throw new Error('Error al obtener productos con stock');
      }
      return await respuesta.json();
    } catch (error) {
      console.error('Error en obtenerConStock:', error);
      throw error;
    }
  }
};

// Exportar el servicio para usarlo en otros archivos
export default productoService;
